maintenance/githubutil/githubutil_test.go

```diff
@@ -0,0 +1,242 @@
+				t.Fatalf("GetCombinedStatus test '%s' failed because the returned value was not a CombinedStatus.", test.name)
+			}
+			if len(combStatus.Statuses) != test.pages {
+				t.Errorf("GetCombinedStatus test '%s' failed because the number of pages returned was %d instead of %d.", test.name, len(combStatus.Statuses), test.pages)
+			}
+			for index, status := range combStatus.Statuses {
+				if status.Context == nil {
+					t.Fatalf("GetCombinedStatus test '%s' failed because the status at index %d had a nil Context field.", test.name, index)
+				}
+				expectedContext := fmt.Sprintf("context %d", index+1)
+				if *status.Context != expectedContext {
+					t.Errorf("GetCombinedStatus test '%s' failed because the status at index %d had a context of '%s' instead of '%s'.", test.name, index, *status.Context, expectedContext)
+				}
+			}
+		},
+	}
+	group.test(t)
+}
+
+// TestForEachPR tests the ForEachPR function to ensure it does retries properly, can handle pagination,
+// and returns the correct result.
+func TestForEachPR(t *testing.T) {
+	processed := 0
+	group := &testGroup{
+		name: "ForEachPR",
+		tests: []*fakeGithub{
+			&fakeGithub{name: "no retries", hitsBeforeResponse: 1, shouldSucceed: true, pages: 1},
+			&fakeGithub{name: "max retries", hitsBeforeResponse: 6, shouldSucceed: true, pages: 1},
+			&fakeGithub{name: "1 too many retries needed", hitsBeforeResponse: 7, shouldSucceed: false, pages: 1},
+			&fakeGithub{name: "3 too many retries needed", hitsBeforeResponse: 10, shouldSucceed: false, pages: 1},
+			&fakeGithub{name: "2 pages", hitsBeforeResponse: 1, shouldSucceed: true, pages: 2},
+			&fakeGithub{name: "10 pages", hitsBeforeResponse: 1, shouldSucceed: true, pages: 10},
+			&fakeGithub{name: "2 pages 2 retries", hitsBeforeResponse: 3, shouldSucceed: true, pages: 2},
+			&fakeGithub{name: "10 pages max retries", hitsBeforeResponse: 6, shouldSucceed: true, pages: 10},
+			&fakeGithub{name: "10 pages one too many retries", hitsBeforeResponse: 7, shouldSucceed: false, pages: 10},
+		},
+		run: func(client *Client) (interface{}, error) {
+			processed = 0
+			process := func(pr *github.PullRequest) error {
+				processed++
+				if pr == nil || pr.Title == nil {
+					return fmt.Errorf("pr %d was invalid", processed)
+				}
+				expectedTitle := fmt.Sprintf("pr %d", processed)
+				if *pr.Title != expectedTitle {
+					return fmt.Errorf("expected pr title '%s' but got '%s'", expectedTitle, *pr.Title)
+				}
+				return nil
+			}
+			return nil, client.ForEachPR("org", "repo", &github.PullRequestListOptions{}, process)
+		},
+		check: func(i interface{}, test *fakeGithub, t *testing.T) {
+			if processed != test.pages {
+				t.Errorf("ForEachPR test '%s' processed %d PRs, but there were only %d single entry pages.", test.name, processed, test.pages)
+			}
+		},
+	}
+	group.test(t)
+}
+
+// testGroup is a struct that runs a group of tests that test the same function and verify it's results in the same way.
+type testGroup struct {
```

Comment:
I would do away with this pattern. I don't think it makes your tests any more readable, and it's a break from the norm.